File: Implementation/2504번 - 괄호의 값/2504.cpp
#include <iostream>
#include <stack>

using namespace std;

string pare;
stack<char> st;

void input() {
    cin >> pare;
}

void solve() {
    int ans = 0, tmp = 1;
    char pre = ' ';
    
    for(int i = 0; i < pare.length(); i++) {
        if(pare[i] == '(') {
            st.push(pare[i]);
            tmp *= 2;
        }
        else if(pare[i] == '[') {
            st.push(pare[i]);
            tmp *= 3;
        }
        else if(pare[i] == ')') {
            if(st.empty() || st.top() == '[') {
                ans = 0;
                break;
            }
            
            st.pop();
            
            if(pre == '(') ans += tmp;
            tmp /= 2;
        }
        else if(pare[i] == ']') {
            if(st.empty() || st.top() == '(') {
                ans = 0;
                break;
            }
            
            st.pop();
            
            if(pre == '[') ans += tmp;
            tmp /= 3;
        }
        
        pre = pare[i];
    }
    
    cout << (st.empty() ? ans : 0) << '\n';
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    input();
    solve();
    
    return 0;
}

File: Implementation/2504번 - 괄호의 값/2504.py
import sys

input = sys.stdin.readline

pare = input().rstrip()
st = []

def solve():
    ans, tmp = 0, 1
    pre = ' '
    for c in pare:
        if c == '(':
            st.append(c)
            tmp *= 2
        elif c == '[':
            st.append(c)
            tmp *= 3
        elif c == ')':
            if not st or st[-1] == '[':
                ans = 0
                break
            
            st.pop()
            
            if pre == '(':
                ans += tmp
            tmp //= 2
            
        elif c == ']':
            if not st or st[-1] == '(':
                ans = 0
                break
            
            st.pop()
            
            if pre == '[':
                ans += tmp
            tmp //= 3
        
        pre = c
    
    print(ans if not st else 0)

if __name__ == '__main__':
    solve()